docs/articles/documentation/recipes/best-practices/best-practices.md

```diff
@@ -0,0 +1,223 @@
+* Clean up between Tests (beforeEach)
+* E2E Tests Limitations
+* Selectors strategy
+
+## Smart Assertions
+
+In end-to-end web testing, peripheral factors (like network lag, processor, or memory bottlenecks in containers) can interfere with the assertions and render tests inconclusive (otherwise known as 'flaky').
+
+Out of the box, TestCafe includes a [Smart Assertion Query Mechanism](../../guides/basic-guides/assert.md#smart-assertion-query-mechanism), which helps mitigate these effects. This mechanism introduces additional wait time for all the assertions: if an assertion fails, the test doesn't fail straight away. Instead, the assertion retries multiple times within a timeout. That helps to stabilize the tests without a big impact on performance.
+
+However, the smart assertions query mechanism only works with promises returned from TestCafe's own `Selector API`.
+
+The following example demonstrates the potential issue:
+
+```js
+import { Selector } from 'testcafe';
+
+fixture `My fixture`
+    .page `http://devexpress.github.io/testcafe/example/`;
+
+test('Assertion with Selector', async t => {
+    const developerNameInput = Selector('#developer-name');
+
+    //this technique leads to unstable results. Avoid it.
+    const developerName      = await Selector('#developer-name').value;
+    //
+
+    console.log(developerName);
+
+    await t
+        .wait(1000)
+        .expect(developerNameInput.value).eql('')
+        .typeText(developerNameInput, 'Peter')
+        .expect(developerName).eql('Peter');
+});
+```
+
+In this snippet, the `developerName` is initialized with the value of a Selector, but because of the `await` keyword, the value is calculated once and doesn't ever update. This effectively disables the smart assertions query mechanism, which in this case leads to inconclusive results. This test fails.
+
+To avoid this, initialize the variable with a `Selector API` promise. To do that, omit the `await` keyword:
+
+```js
+import { Selector } from 'testcafe';
+
+fixture `My fixture`
+    .page `http://devexpress.github.io/testcafe/example/`;
+
+test('Assertion with Selector', async t => {
+    const developerNameInput = Selector('#developer-name');
+    const developerName      = Selector('#developer-name').value;
+
+    console.log(developerName);
+
+    await t
+        .wait(1000)
+        .expect(developerNameInput.value).eql('')
+        .typeText(developerNameInput, 'Peter')
+        .expect(developerName).eql('Peter');
+});
+```
+
+The `developerName` is now initialized with a re-executable Selector API promise. When a test controller receives this promise, it enables the smart assertion query mechanism and TestCafe can wait for the value to update.
+
+> The smart assertion query mechanism works with promises generated by TestCafe Selector API. Use the `await` keyword with user-created promises and promises generated by third-party libraries.
+
+This rule applies to ClientFunctions in the same capacity.
```

Comment:
> in the same capacity.

Same here, I don't understand this expression. It's either too rare (we shouldn't use it in this case) or it's just me (then if the correctors approve we can use it)

Reply:
This rule applies to ClientFunctions **as well**.
OR
This rule applies to ClientFunctions.